lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs

```diff
@@ -0,0 +1,511 @@
+        } else {
+            Err(custom_err!(
+                "FilesystemBackend::handle_tdp_sd_info_response",
+                FilesystemBackendError(format!(
+                    "received invalid completion id: {}",
+                    res.completion_id
+                ))
+            ))
+        }
+    }
+
+    /// Helper function for sending a TDP SharedDirectoryCreateRequest based on an
+    /// RDP DeviceCreateRequest and handling the TDP SharedDirectoryCreateResponse.
+    fn tdp_sd_create(
+        &mut self,
+        rdp_req: efs::DeviceCreateRequest,
+        file_type: tdp::FileType,
+    ) -> PduResult<()> {
+        todo!()
+    }
+
+    /// Helper function for combining a TDP SharedDirectoryDeleteRequest
+    /// with a TDP SharedDirectoryCreateRequest to overwrite a file, based
+    /// on an RDP DeviceCreateRequest.
+    fn tdp_sd_overwrite(&mut self, rdp_req: efs::DeviceCreateRequest) -> PduResult<()> {
+        todo!()
+    }
+
+    fn make_device_create_response(
+        device_create_request: &efs::DeviceCreateRequest,
+        io_status: efs::NtStatus,
+        new_file_id: u32,
+    ) -> PduResult<Option<RdpdrPdu>> {
+        // See https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_main.c#L187-L228
+        let information = if io_status != efs::NtStatus::Success
+            || device_create_request.create_disposition.intersects(
+                efs::CreateDisposition::FILE_SUPERSEDE
+                    | efs::CreateDisposition::FILE_OPEN
+                    | efs::CreateDisposition::FILE_CREATE
+                    | efs::CreateDisposition::FILE_OVERWRITE,
+            ) {
+            Ok(efs::Information::FILE_SUPERSEDED)
+        } else if device_create_request.create_disposition == efs::CreateDisposition::FILE_OPEN_IF {
+            Ok(efs::Information::FILE_OPENED)
+        } else if device_create_request.create_disposition
+            == efs::CreateDisposition::FILE_OVERWRITE_IF
+        {
+            Ok(efs::Information::FILE_OVERWRITTEN)
+        } else {
+            Err(other_err!(
+                "FilesystemBackend::make_device_create_response",
+                "program error, CreateDispositionFlags check should be exhaustive"
+            ))
+        }?;
+
+        Ok(Some(RdpdrPdu::DeviceCreateResponse(
+            efs::DeviceCreateResponse {
+                device_io_reply: efs::DeviceIoResponse::new(
+                    device_create_request.device_io_request,
+                    io_status,
+                ),
+                file_id: new_file_id,
+                information,
+            },
+        )))
+    }
+}
+
+#[derive(Debug)]
+struct FileCache {
+    cache: HashMap<u32, FileCacheObject>,
+    next_file_id: u32,
+}
+
+impl FileCache {
+    fn new() -> Self {
+        Self {
+            cache: HashMap::new(),
+            next_file_id: 0,
+        }
+    }
+
+    /// Insert a [`FileCacheObject`] into the file cache.
```

Comment:
Is the square bracket syntax a rustdoc thing? Why do we only use it on the doc for `insert` but not for `get`, `get_mut`, or `remove`?

Reply:
Yeah that's a rustdoc thing: https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html

I added it for the other methods in c5cdbed27388296f01f2071c327f0bee975c5676